Dockerfile

```diff
@@ -192,20 +175,60 @@ ENV _ASH_EXEC_MODE="local"
 #
 ENV PATH="$PATH:/ash"
 
-# nosemgrep
-HEALTHCHECK --interval=12s --timeout=12s --start-period=30s \
-    CMD type ash || exit 1
+
+# CI stage -- any customizations specific to CI platform compatibility should be added
+# in this stage if it is not applicable to ASH outside of CI usage
+FROM core as ci
+
+ENV ASH_TARGET=ci
+
+
+# Final stage: Non-root user final version. This image contains all dependencies
+# for ASH from the `core` stage, but ensures it is launched as a non-root user.
+# Running as a non-root user impacts the ability to run ASH reliably across CI
+# platforms and other orchestrators where the initialization and launch of the image
+# is not configurable for customizing the running UID/GID.
+FROM core as non-root
+
+ENV ASH_TARGET=non-root
+
+ARG UID=500
+ARG GID=100
+ARG ASH_USER=ash-user
+ARG ASH_GROUP=ash-group
+ARG ASHUSER_HOME=/home/${ASH_USER}
 
 #
-# The ENTRYPOINT needs to be NULL for CI platform support
-# This needs to be an empty array ([ ]), as nerdctl-based runners will attempt to
-# resolve an empty string in PATH, unlike Docker which treats an empty string the
-# same as a literal NULL
+# Create a non-root user in the container and run as this user
 #
-ENTRYPOINT [ ]
+# And add GitHub's public fingerprints to known_hosts inside the image to prevent fingerprint
+# confirmation requests unexpectedly
+#
+# ignore a failure to add the group
+RUN addgroup --gid ${GID} ${ASH_GROUP} || :
+RUN adduser --disabled-password --disabled-login \
+        --uid ${UID} --gid ${GID} \
+        ${ASH_USER} && \
+    mkdir -p ${ASHUSER_HOME}/.ssh && \
+    echo "github.com ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOMqqnkVzrm0SdG6UOoqKLsabgH5C9okWi0dh2l9GKJl" >> ${ASHUSER_HOME}/.ssh/known_hosts && \
+    echo "github.com ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEmKSENjQEezOmxkZMy7opKgwFB9nkt5YRrYMjNuG5N87uRgg6CLrbo5wAdT/y6v0mKV0U2w0WZ2YB/++Tpockg=" >> ${ASHUSER_HOME}/.ssh/known_hosts && \
+    echo "github.com ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQCj7ndNxQowgcQnjshcLrqPEiiphnt+VTTvDP6mHBL9j1aNUkY4Ue1gvwnGLVlOhGeYrnZaMgRK6+PKCUXaDbC7qtbW8gIkhL7aGCsOr/C56SJMy/BCZfxd1nWzAOxSDPgVsmerOBYfNqltV9/hWCqBywINIR+5dIg6JTJ72pcEpEjcYgXkE2YEFXV1JHnsKgbLWNlhScqb2UmyRkQyytRLtL+38TGxkxCflmO+5Z8CSSNY7GidjMIZ7Q4zMjA2n1nGrlTDkzwDCsw+wqFPGQA179cnfGWOWRVruj16z6XyvxvjJwbz0wQZ75XK5tKSb7FNyeIEs4TT4jk+S4dhPeAUC5y+bDYirYgM4GC7uEnztnZyaVWQ7B381AK4Qdrwt51ZqExKbQpTUNn+EjqoTwvqNj4kqx5QUCI0ThS/YkOxJCXmPUWZbhjpCg56i+2aB6CmK2JGhn57K5mj0MNdBXA4/WnwH6XoPWJzK5Nyu2zB3nAZp+S5hpQs+p1vN1/wsjk=" >> ${ASHUSER_HOME}/.ssh/known_hosts && \
+    chown -R ${UID}:${GID} ${ASHUSER_HOME}
+# Setting default WORKDIR to ${ASHUSER_HOME}
```

Comment:
Now that we know the UID/GID value - should we `chown` and `chgrp` `/out` and `/deps` here?

Reply:
Yep, I think we can further restrict and take ownership here since this becomes known at this point. We still need the earlier `chmod` otherwise the `ci` stage built images do not work in Azure Pipelines due to permissions issues.

Reply:
ok

Reply:
![CleanShot 2024-10-31 at 21 25 55](https://github.com/user-attachments/assets/eefcee74-c573-45c0-9093-ae4fd6022690)

Reply:
Thoughts here, @climbertjh2 ? Let me know if this should be good to resolve.

Reply:
@scrthq  - my read of the `chown` docs is that if you `chown <UID>:<GID>` then you don't need to `chgrp <GID>`.

So the second line/statement in the RUN block is not needed.

Reply:
Other than that, it looks good to me.

Reply:
10-4, removing.